cascaritaco/cascarita
# Feature/create user log in and registration endpoints

**### WIP WORKING ON ADDING TESTS** 
**Summary**

1. Created User, Role and Group Models. Included migration create the tables in the database. May notice that the User migration contains explicit associations between the Group and Role tables as `groupId` and `roleId` are foreign keys in the user table 
2. Created controllers for each model, with simple create functionality 
3. User model contains functionality to hash and salt user's passwords when creating a new password using the `Bcrypt` package. Also uses it to compare passwords when user logs in. This hashing and compare passwords logic is in the user model file but its used in the `config/passport.js` file. 
4. Cleaned up `app.js` by using router, thus created `router/modelName.routes.js` file system. 
5.  Updates: Using `validate()` to validate data against the model's constraints !! And all wrapped in the `try` `catch` so if there is an error it won't crash the app

Manual Testing:

1.  if in `main` branch, `git pull`
2.  switch onto the branch `git checkout feature/craete-user-log-in-n-registration `
3. run `npm install`
4. run migration so you get the new User, Group and Role tables !!!!!:  If in root directory then `cd server` as there is where the migration folder is located in and then run `npx sequelize-cli db:migrate`
5. Double check your database and make sure new migrations ran successfully and you have the new tables 
6. Go to post man and  set to send a `post` request
7. `http://localhost:80/[model-name]/[end-point-that-can-be-found-in-the-route-files-in-the-route-folder` 
8. go to the **Body** tab 
9. select `x-www-form-urlencoded`
10.  set up key value pairs. Reference the model to get the names of the keys. Note if you want to create a new user like in the screenshot provided there needs to be an existing group and role because you need to include a `groupId` and `roleId` 
<img width="1378" alt="Screenshot 2023-12-13 at 3 45 49 PM" src="https://github.com/cascaritaco/cascarita/assets/43281985/ff16e9b1-3d23-4197-8169-3004114aca04">
